modules/python3-cryptography.yml

```diff
@@ -0,0 +1,20 @@
+name: python3-cryptography
+buildsystem: simple
+build-commands:
+  - pip3 install --verbose --no-index --find-links="file://${PWD}" --prefix=${FLATPAK_DEST} --no-build-isolation
+    cryptography
+sources:
+  - type: file
+    url: https://files.pythonhosted.org/packages/68/ce/95b0bae7968c65473e1298efb042e10cafc7bafc14d9e4f154008241c91d/cffi-1.16.0.tar.gz
+    sha256: bcb3ef43e58665bbda2fb198698fcae6776483e0c4a631aa5647806c25e02cc0
+  - type: file
+    url: https://files.pythonhosted.org/packages/d4/fa/057f9d7a5364c86ccb6a4bd4e5c58920dcb66532be0cc21da3f9c7617ec3/cryptography-42.0.5-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```

Comment:
is this binary really compatible? It's python 3.11 in that runtime.

Reply:
This should be built from source.

Reply:
According to https://pypi.org/project/cryptography/#modal-close there are noe packages for Python 3.11.

Reply:
> According to https://pypi.org/project/cryptography/#modal-close there are noe packages for Python 3.11.

It should be built from source. You can download the tar.gz from pypi extract it, go to the src/cargo folder and use the flatpak-cargo-generator on the cargo.lock. This will generate a cargo dependency manifest. Then add that and add rust-stable to the manifest here.

See for example https://github.com/flathub/io.github.hydrusnetwork.hydrus/blob/master/python3-sources.yaml#L591-L610

Reply:
I added the SDK, extracted the cargo sources as you described and now get this:

```
 $ flatpak run de.nagstamon.nagstamon

Traceback (most recent call last):
  File "/app/lib/python3.11/site-packages/Nagstamon/Config.py", line 710, in KeyringAvailable
    import secretstorage
  File "/app/lib/python3.11/site-packages/secretstorage/__init__.py", line 13, in <module>
    from secretstorage.collection import Collection, create_collection, \
  File "/app/lib/python3.11/site-packages/secretstorage/collection.py", line 24, in <module>
    from secretstorage.item import Item
  File "/app/lib/python3.11/site-packages/secretstorage/item.py", line 17, in <module>
    from secretstorage.util import DBusAddressWrapper, \
  File "/app/lib/python3.11/site-packages/secretstorage/util.py", line 23, in <module>
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
    from cryptography.hazmat.primitives.ciphers.base import (
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
    from cryptography.exceptions import (
  File "/app/lib/python3.11/site-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
ImportError: cannot import name 'exceptions' from 'cryptography.hazmat.bindings._rust' (unknown location)
Traceback (most recent call last):
  File "/app/lib/python3.11/site-packages/Nagstamon/Config.py", line 710, in KeyringAvailable
    import secretstorage
  File "/app/lib/python3.11/site-packages/secretstorage/__init__.py", line 13, in <module>
    from secretstorage.collection import Collection, create_collection, \
  File "/app/lib/python3.11/site-packages/secretstorage/collection.py", line 24, in <module>
    from secretstorage.item import Item
  File "/app/lib/python3.11/site-packages/secretstorage/item.py", line 17, in <module>
    from secretstorage.util import DBusAddressWrapper, \
  File "/app/lib/python3.11/site-packages/secretstorage/util.py", line 23, in <module>
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
    from cryptography.hazmat.primitives.ciphers.base import (
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
    from cryptography.exceptions import (
  File "/app/lib/python3.11/site-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
ImportError: cannot import name 'exceptions' from 'cryptography.hazmat.bindings._rust' (unknown location)
Traceback (most recent call last):
  File "/app/lib/python3.11/site-packages/Nagstamon/Config.py", line 710, in KeyringAvailable
    import secretstorage
  File "/app/lib/python3.11/site-packages/secretstorage/__init__.py", line 13, in <module>
    from secretstorage.collection import Collection, create_collection, \
  File "/app/lib/python3.11/site-packages/secretstorage/collection.py", line 24, in <module>
    from secretstorage.item import Item
  File "/app/lib/python3.11/site-packages/secretstorage/item.py", line 17, in <module>
    from secretstorage.util import DBusAddressWrapper, \
  File "/app/lib/python3.11/site-packages/secretstorage/util.py", line 23, in <module>
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
    from cryptography.hazmat.primitives.ciphers.base import (
  File "/app/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
    from cryptography.exceptions import (
  File "/app/lib/python3.11/site-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
ImportError: cannot import name 'exceptions' from 'cryptography.hazmat.bindings._rust' (unknown location)
cannot import name 'exceptions' from 'cryptography.hazmat.bindings._rust' (unknown location)
cannot import name 'exceptions' from 'cryptography.hazmat.bindings._rust' (unknown location)
```

Do you have any hint? Or is there an option to use the precompiled wheel of cryptography?

Reply:
Can you try adding setuptools_rust before cffi? https://github.com/flathub/io.github.hydrusnetwork.hydrus/blob/0365ee55448838d93a96659a29a89bc509b709bf/python3-sources.yaml#L548-L569 Pretty sure it's skipping building some bindings due to lack of a dependency.

Reply:
@bbhtt thanks, this was the solution! :+1: